src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -18,22 +20,75 @@ namespace AzureMcp.Services.Azure.Authentication;
+        return _chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+    }
 
-    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken) =>
-        _chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+    private const string AuthenticationRecordEnvVarName = "AZURE_MCP_AUTHENTICATION_RECORD";
+    private const string OnlyUseBrokerCredentialEnvVarName = "AZURE_MCP_ONLY_USE_BROKER_CREDENTIAL";
+    private const string ClientIdEnvVarName = "AZURE_MCP_CLIENT_ID";
 
-    private static ChainedTokenCredential CreateChainedCredential(string? tenantId) =>
-        new(CreateDefaultCredential(tenantId), CreateBrowserCredential());
+    private static bool ShouldUseOnlyBrokerCredential()
+    {
+        return EnvironmentHelpers.GetEnvironmentVariableAsBool(OnlyUseBrokerCredentialEnvVarName);
+     }
 
-    private static InteractiveBrowserCredential CreateBrowserCredential() =>
-        new(new InteractiveBrowserCredentialBrokerOptions(IntPtr.Zero)
+    private static ChainedTokenCredential CreateChainedCredential(string? tenantId)
+    {
+        string? authRecordJson = Environment.GetEnvironmentVariable(AuthenticationRecordEnvVarName);
+        AuthenticationRecord? authRecord = null;
+        if (!string.IsNullOrEmpty(authRecordJson))
         {
-            UseDefaultBrokerAccount = true
-        });
+            byte[] bytes = Encoding.UTF8.GetBytes(authRecordJson);
+            using MemoryStream authRecordStream = new MemoryStream(bytes);
+            authRecord = AuthenticationRecord.Deserialize(authRecordStream);
+        }
+
+        if (ShouldUseOnlyBrokerCredential())
+        {
+            return new(CreateBrowserCredential(tenantId, authRecord));
+        }
+        else
+        {
+            return new(CreateDefaultCredential(tenantId), CreateBrowserCredential(tenantId, authRecord));
+        }
+    }
+
+    private static string TokenCacheName = "azure-mcp-msal.cache";
+
+    private static InteractiveBrowserCredential CreateBrowserCredential(string? tenantId, AuthenticationRecord? authRecord)
+    {
+        string? clientId = Environment.GetEnvironmentVariable(ClientIdEnvVarName);
+
+        IntPtr handle = WindowHandleProvider.GetWindowHandle();
+
+        InteractiveBrowserCredentialBrokerOptions brokerOptions = new(handle)
+        {
+            UseDefaultBrokerAccount = !ShouldUseOnlyBrokerCredential() && authRecord is null,
+            TenantId = tenantId,
+            AuthenticationRecord = authRecord,
+            TokenCachePersistenceOptions = new TokenCachePersistenceOptions()
+            {
+                Name = TokenCacheName,
+            }
+        };
+        if (clientId is not null)
+        {
+            brokerOptions.ClientId = clientId;
+        }
+
+        return new(brokerOptions);
+    }
 
     private static DefaultAzureCredential CreateDefaultCredential(string? tenantId)
     {
```

Comment:
Use the new truthy helper

Reply:
If you are referring to the clientId, its logic is different from the logic of the truth helper.

Reply:
![image](https://github.com/user-attachments/assets/1f5730b7-fb54-4295-9d9f-269f0763983c)